### PR TITLE
[Workflow] Example on reacting to successful transitions

### PR DESCRIPTION
**Pull request by kix**
This PR broadens Workflow documentation, explaining which events are dispatched during workflow transitions. @lyrixx, can you clarify whether these events are a part of the component API? Reacting to a successful transition is a quite common scenario, IMHO.

### CODE REVIEW

File: workflow/usage.rst [hunk 1]
```diff
@@ -175,12 +175,63 @@ what actions are allowed on a blog post::
...
-object with an ``EventDispatcher``. You can now create event listeners to
-block transitions (i.e. depending on the data in the blog post). The following
-events are dispatched:
+object with an ``EventDispatcher``. If you are using the ``Workflow`` component within
+the Symfony standard edition, this is already done for you.
+
+When a workflow subject's state is updated, the component dispatches several events that
+enable you to alter the component's behaviour:
+
+* ``workflow.[workflow name].guard.[transition name]``: dispatched before applying a transition.
```
Comment (lyrixx) [inline workflow/usage.rst]: Actually, this event is a bit special because it allows to "block" a transition with pure PHP code.

File: workflow/usage.rst [hunk 2]
```diff
@@ -175,12 +175,63 @@ what actions are allowed on a blog post::
...
+* ``workflow.[workflow name].leave.[state name]``: a subject leaves a state.
+* ``workflow.[workflow name].transition.[transition name]``: a transition is being applied.
+* ``workflow.[workflow name].enter.[state name]``: a subject enters a state.
+* ``workflow.[workflow name].announce.[transition name]``: a transition was applied to the subject.
```
Comment (kix) [inline workflow/usage.rst]: Perhaps this deserves a note that a `GuardEvent` will be dispatched for guards instead of a base `Workflow\Event`.

File: workflow/usage.rst (hunk 1 above)
Comment (lyrixx) [inline workflow/usage.rst]: Each time an "event" is dispatched, there are actually 3 events:
* `workflow.guard`
* `workflow.[workflow name].guard`
* `workflow.[workflow name].guard.[transition name]`

cf https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L217-L219

The same comments applies to others event too.

File: workflow/usage.rst [hunk 3]
```diff
@@ -175,12 +175,63 @@ what actions are allowed on a blog post::
...
+                'workflow.blogpost.announce.to_review' => array('notifyPostProposal'),
+            );
+        }
+    }
+
+Blocking transitions
+--------------------
+
+You can create Guard event listeners to block transitions (i.e. depending on the data in the blog
+post). The following events are dispatched:
```
Comment (kix) [inline workflow/usage.rst]: The events have already been listed in the previous section here. We could remove the list here, and add a note on `GuardEvent` here.

File: workflow/usage.rst [hunk 4]
```diff
@@ -175,12 +175,63 @@ what actions are allowed on a blog post::
...
+
+        public function notifyPostProposal(Event $event)
+        {
+            /** @var \AppBundle\Entity\BlogPost $post */
+            $post = $event->getSubject();
+
+            $message = \Swift_Message::newInstance('New blog post proposed');
+            $message->setTo(['admin@acme.co' => 'Admin']);
+            $message->setBody('A post has been submitted to review');
+        }
```
Comment (lyrixx) [inline workflow/usage.rst]: You forgot to really send the mail.

File: workflow/usage.rst (hunk 1 above)
Comment (kix) [inline workflow/usage.rst]: I know, but that's a thing that feels hard to explain if using the format given.

File: workflow/usage.rst (hunk 4 above)
Comment (kix) [inline workflow/usage.rst]: Fixed :)

File: workflow/usage.rst (hunk 1 above)
Comment (kix) [inline workflow/usage.rst]: I could reword this as:
`workflow.[workflow name].guard.[transition name]`: dispatched before applying a transition. Can be used to block transitions from listeners.

Comment (lyrixx) [inline workflow/usage.rst]: I should only be use to block (or not). Not for something else ;) But as I'm not a native english speaker I can not tell what's the best way to say it.

Comment (kix) [inline workflow/usage.rst]: Not a native speaker as well, but let's maybe wait for @javiereguiluz's review? I guess he'll come up with something here :) Not quite sure about the best wording.

Comment (kix) [inline workflow/usage.rst]: Actually, I'm not sure what's the point of the generic `workflow.guard`/`leave`/`transition` etc. events. Logging of each and every update for all available workflows?
Same for events having a `[workflow name]` (e.g. `workflow.blogpost.guard`), the only use case I can see here is logging. Guarding every step in a workflow can only be useful for cases when the workflow can be started for a subject that's in an unacceptable state.
Hopefully, we can clarify this a bit?

Comment (kix) [inline workflow/usage.rst]: Reworded as `used to conditionally block a transition`

File: workflow/usage.rst [hunk 5]
```diff
@@ -175,12 +175,65 @@ what actions are allowed on a blog post::
...
-object with an ``EventDispatcher``. You can now create event listeners to
-block transitions (i.e. depending on the data in the blog post). The following
-events are dispatched:
+object with an ``EventDispatcher``. If you are using the ``Workflow`` component within
+the Symfony standard edition, this is already done for you.
+
+When a workflow subject's state is updated, the component dispatches several events that
+enable you to alter the component's behaviour:
+
+* ``workflow.[workflow name].guard.[transition name]``: used to conditionally block a transition.
```
Comment (javiereguiluz) [inline workflow/usage.rst]: I think we should reword these descriptions a bit. The essential thing to document about events is if they are notified before or after something happened (that's why it's recommended to name the events `post_*` and `pre_*`). For example, `a transition is being applied` could be be: `notified just before applying a transition`

**Comment by HeahDude** [comment]
This is now documented, http://symfony.com/doc/current/workflow/usage.html. Shall we close?

**Comment by xabbuh** [comment]
I agree with @HeahDude here.

@kix Please let us know if you don't agree or if you have other ideas how to improve the documentation here.

**Comment by kix** [comment]
@xabbuh, looks good to me! Thanks!
